orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/dynomite/DynomiteExecutionRepository.java

```diff
@@ -0,0 +1,265 @@
+          execution.getId()
+        );
+      }
+    });
+  }
+
+  private void storeStageInternal(RedisClientDelegate delegate, Stage stage, Boolean updateIndex) {
+    String key = executionKey(stage);
+    String indexKey = format("%s:stageIndex", key);
+
+    Map<String, String> serializedStage = serializeStage(stage);
+    List<String> keysToRemove = serializedStage.entrySet().stream()
+      .filter(e -> e.getValue() == null)
+      .map(Map.Entry::getKey)
+      .collect(Collectors.toList());
+
+    serializedStage.values().removeIf(Objects::isNull);
+
+    delegate.withPipeline(pipeline -> {
+      DynoJedisPipeline p = (DynoJedisPipeline) pipeline;
+      p.hmset(key, serializedStage);
+
+      if (!keysToRemove.isEmpty()) {
+        p.hdel(key, keysToRemove.toArray(new String[0]));
+      }
+      if (updateIndex) {
+        p.sadd(indexKey, stage.getId());
+      }
+      p.sync();
+    });
+  }
+
+  @Override
+  protected Execution retrieveInternal(RedisClientDelegate redisClientDelegate, ExecutionType type, String id) throws ExecutionNotFoundException {
+    String key = format("{%s:%s}", type, id);
+    String indexKey = format("%s:stageIndex", key);
+
+    boolean exists = redisClientDelegate.withCommandsClient(c -> {
+      return c.exists(key);
+    });
+    if (!exists) {
+      throw new ExecutionNotFoundException("No ${type} found for $id");
+    }
+
+    final Map<String, String> map = new HashMap<>();
+    final List<String> stageIds = new ArrayList<>();
+
+    redisClientDelegate.withPipeline(pipeline -> {
+      DynoJedisPipeline p = (DynoJedisPipeline) pipeline;
+      Response<Map<String, String>> execResponse = p.hgetAll(key);
+      Response<Set<String>> indexResponse = p.smembers(indexKey);
+      try {
+        p.sync();
+      } catch (Exception e) {
+        log.error(e.getMessage(), e);
+      }
+
+      map.putAll(execResponse.get());
+
+      if (!indexResponse.get().isEmpty()) {
+        stageIds.addAll(indexResponse.get());
+      } else {
+        stageIds.addAll(extractStages(map));
+      }
+    });
+
+    Execution execution = new Execution(type, id, map.get("application"));
+    return buildExecution(execution, map, stageIds);
+  }
+
+  @Override
+  protected List<String> fetchMultiExecutionStatus(RedisClientDelegate redisClientDelegate, List<String> keys) {
+    List<String> statuses = new ArrayList<>();
+    redisClientDelegate.withCommandsClient(c -> {
+      keys.forEach(k -> statuses.add(c.hget(k, "status")));
+    });
+    return statuses;
+  }
+
+  @Override
+  protected String executionKey(Execution execution) {
+    return format("{%s:%s}", execution.getType(), execution.getId());
+  }
+
+  @Override
+  protected String executionKey(Stage stage) {
+    return format("{%s:%s}", stage.getExecution().getType(), stage.getExecution().getId());
+  }
+
+  @Override
+  protected String pipelineKey(String id) {
+    return format("{pipeline:%s}", id);
```

Comment:
I think this and `orchestrationKey` should use the `ExecutionType` enum and `toString` instead.

Reply:
➕